### PR TITLE
added dockerfile for pangolin 3.1.20 using micromamba as base in an a…

### PR DESCRIPTION
**Pull request by kapsakcj**
…ttempt to get a smaller docker image

- [x] This comment contains a description of what is in the pull request.

- [x] Build your own docker image using a Dockerfile
  - [x] Directory structure should be name of the tool in lower case with special characters removed with a subdirectory of the version number (i.e. spades/3.12.0/Dockerfile)
  - [x] Includes the recommended LABELS
- [x] (Optional) Dockerfile is built with best practices and has been approved by a linter (such as https://hadolint.github.io/hadolint/)
- [x] Edit main README.md
- [x] Ensure tool is listed in Program_Licenses.md
- [x] Create a simple container-specific README.md in the same directory as the Dockerfile (i.e. spades/3.12.0/README.md)
- [x] Update GitHub actions workflow if needed
- [x] Any files required for building are located in the same directory as the Dockerfile (i.e. spades/3.12.0/my_spades_tests.sh)
- [ ] Have successfully run the workflow "Test <program name> image" in your forked repository

### CODE REVIEW

**Comment by kapsakcj** [comment]
The above dockerfile builds into an image successfully and is a little bit smaller than the current docker image for pangolin 3.1.19 & pangolearn 2022-01-20

It ran successfully on some Omicron genomes and the test_seqs included with pangolin.

There is about a 640 MB difference in uncompressed docker image sizes:
```
$ docker images | grep '3.1.19'
kapsakcj/pangolin                                                    3.1.19-slim                            4a4fddec0534   About an hour ago   2.15GB
staphb/pangolin                                                      3.1.19-pangolearn-2022-01-20           e867e547cc11   3 days ago          2.79GB
```

**Comment by kapsakcj** [comment]
Keeping as a draft for now until I polish up the dockerfile

TODO - test with singularity

**Comment by erinyoung** [comment]
That only shaved off half a gig of space. Do you know what's taking up the most room in the container?

**Comment by kapsakcj** [comment]
I plan on updating this to the latest versions of pangolin/pangolearn/constellations/etc. Probably will offer this alongside the usual docker image. Might call the docker image `staphb/pangolin:3.1.20-pangolearn-2022-02-02-slim` to differentiate

**Comment by kapsakcj** [comment]
Alrighty, I:
  * upgraded the base image to the most recent micromamba version 
  * upgraded pangolin & dependencies to latest versions
  * Cleaned up comments/formatting of dockerfile.

Tested this alongside the usual install of pangolin (3.1.20 & PL 2022-02-02) and it produced identical outputs on a variety of Omicron genomes & the test sequences supplied w/ pangolin.

I think this docker image is ready for deployment and I propose we tag it as `staphb/pangolin:3.1.20-pangolearn-2022-02-02-slim`

Does anyone want to test before I merge and build on dockerhub/quay? @k-florek @kevinlibuit @erinyoung

**Comment by kevinlibuit** [comment]
Giving it a review now!
- [ ] successful build 
- [ ] local test

**Comment by kevinlibuit** [comment]
Details on review:
<details>
 <summary>Successful build</summary>
 Built successfully on mid-size vm (4 vCPUs, 16 GB memory):
 
```
$  time docker build --tag cjk/pangolin:1.2.20-slim ~/docker-builds/pangolin/3.1.20-slim/

Sending build context to Docker daemon   5.12kB
Step 1/28 : FROM mambaorg/micromamba:0.21.0


[...]

Successfully built a7132c393c6e
Successfully tagged cjk/pangolin:1.2.20-slim

real    3m48.538s
user    0m0.085s
sys     0m0.077s
```
</details>

<details>
 <summary>Local Tests</summary>
 Ran without issue with default params
 
```
$ docker run --rm=True -ti -v $PWD:/data cjk/pangolin:1.2.20-slim
$ pangolin --all-versions
pangolin: 3.1.20
pangolearn: 2022-02-02
constellations: v0.1.3
scorpio: 0.3.16
pango-designation used by pangoLEARN/Usher: v1.2.124
pango-designation aliases: 1.2.127
$ pangolin ./Sample1.medaka.consensus.fasta
All dependencies satisfied.
The query file is:/data/./Sample1.medaka.consensus.fasta
** Running sequence QC **
Number of sequences detected: 1
Total passing QC: 1

Data files found:
Trained model:  /opt/conda/envs/pangolin/lib/python3.8/site-packages/pangoLEARN/data/decisionTree_v1.joblib
Header file:    /opt/conda/envs/pangolin/lib/python3.8/site-packages/pangoLEARN/data/decisionTreeHeaders_v1.joblib
Designated hash:        /opt/conda/envs/pangolin/lib/python3.8/site-packages/pangoLEARN/data/lineages.hash.csv
Job stats:
job                     count    min threads    max threads
--------------------  -------  -------------  -------------
add_failed_seqs             1              1              1
align_to_reference          1              1              1
all                         1              1              1
generate_report             1              1              1
get_constellations          1              1              1
hash_sequence_assign        1              1              1
pangolearn                  1              1              1
scorpio                     1              1              1
total                       8              1              1

loading model 02/11/2022, 19:31:48
/opt/conda/envs/pangolin/lib/python3.8/site-packages/sklearn/base.py:329: UserWarning: Trying to unpickle estimator DecisionTreeClassifier from version 0.24.2 when using version 1.0.2. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:
https://scikit-learn.org/stable/modules/model_persistence.html#security-maintainability-limitations
  warnings.warn(
processing block of 1 sequences 02/11/2022, 19:31:50
/opt/conda/envs/pangolin/lib/python3.8/site-packages/sklearn/base.py:443: UserWarning: X has feature names, but DecisionTreeClassifier was fitted without feature names
  warnings.warn(
complete 02/11/2022, 19:31:50
Output file written to: /data/lineage_report.csv

```

</details>

**Comment by kevinlibuit** [comment]
And while use of micromamba _only_ shaves off 640MB of space, that's >20% reduction in size which is pretty impressive. Awesome job, Curtis! 

When compressed, it seems that most StaPH-B images with conda environments float around that ~1GB range. So this is right in-line with other images in the repo. 

@erinyoung curious to know if you have suggestions to further reduce this image size? And also interested in knowing why this might be an issue?

**Comment by kevinlibuit** [comment]
Passed all tests on my end, but we can wait to merge if @erinyoung would like to try and shrink this any further.

File: pangolin/3.1.20-slim/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ARG MAMBA_DOCKERFILE_ACTIVATE=1
+
+WORKDIR /pangolin
+RUN pip install . && \
+ micromamba clean -a -y && \
+ mkdir /data && \
+ pangolin --all-versions && \
+ usher --version
+
+WORKDIR /data
```
Comment (erinyoung) [inline pangolin/3.1.20-slim/Dockerfile:67]: And then something like this at the end for testing:
```
# testing
FROM app as test

ARG MAMBA_DOCKERFILE_ACTIVATE=1

RUN mkdir /tests
WORKDIR /tests

RUN wget https://raw.githubusercontent.com/StaPH-B/docker-builds/master/tests/SARS-CoV-2/SRR13957123.consensus.fa && pangolin SRR13957123.consensus.fa
```

I know there are test files in the github repo (that's where the wget is pulling files from), but using dockerfile's COPY function only works on files that in the repo or below it. You might find something more elegant.

And then you'd want a corresponding testing file like:
```
name: Test pangolin image

on:
  workflow_dispatch:
  pull_request:
    paths:
      - 'pangolin/*/Dockerfile'

jobs:
  build-to-test:
    uses: ./.github/workflows/build-to-test.yml
    with:
      path_to_context: "pangolin/<whatever this version was>"
      dockerfile_name: "Dockerfile"
      cache: "pangolin"
```

**Review by erinyoung — CHANGES_REQUESTED** [review]
Once you add a test, you'll want a corresponding `.github/workflows/test-pangolin.yml` file:
```
name: Test pangolin image

on:
  workflow_dispatch:
  pull_request:
    paths:
      - 'pangolin/*/Dockerfile'

jobs:
  build-to-test:
    uses: ./.github/workflows/build-to-test.yml
    with:
      path_to_context: "pangolin/<whatever this version was>"
      dockerfile_name: "Dockerfile"
      cache: "pangolin"
```

**Comment by kapsakcj** [comment]
also closing in favor of #311

**Pull request closed without merging**